### PR TITLE
[pytest] Test the feature of memory checker.

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
This PR aims to test the feature of memory checker in 20191130.73 and newer image version. Currently this pytest script only tests the feature on streaming telemetry container and will extend this test to all other containers once this feature is fully implemented.

#### How did you do it?
I implemented this test according to the following steps:
1. Installs `stress` utility in container under testing.
2. Invokes `stress` utility to eat up the amount of memory which is larger than the threshold.
3. Waits for the container to be stopped and restarted
4. Checks whether all critical processes in container are running

#### How did you verify/test it?
I tested this feature on the testbed `str-msn2700-01`.

    18:30:05 test_memory_checker.modify_monit_config_ L0038 INFO | Back up Monit configuration file ...
    18:30:06 test_memory_checker.modify_monit_config_ L0041 INFO | Modifying Monit config to eliminate start delay and decrease interval ...
    18:30:07 test_memory_checker.modify_monit_config_ L0045 INFO | Restart Monit service ...
    18:30:10 __init__.loganalyzer L0047 INFO | Log analyzer is disabled
    -------------------------------- live log call ---------------------------------
    18:30:10 test_memory_checker.install_stress_utili L0067 INFO | Installing 'stress' utility in 'telemetry' container ...
    18:30:15 test_memory_checker.install_stress_utili L0075 INFO | 'stress' utility was installed.
    18:30:15 test_memory_checker.consume_memory_and_r L0129 INFO | Waiting 'telemetry' container to be stopped ...
    18:30:15 test_memory_checker.consume_memory L0107 INFO | Executing command 'stress -m 4' in 'telemetry' container ...
    18:32:27 test_memory_checker.consume_memory_and_r L0134 INFO | 'telemetry' container is stopped.
    18:32:27 test_memory_checker.consume_memory_and_r L0136 INFO | Waiting 'telemetry' container to be restarted ...
    18:32:30 test_memory_checker.consume_memory_and_r L0141 INFO | 'telemetry' container is restarted.
    18:32:30 test_memory_checker.remove_stress_utilit L0088 INFO | Removing 'stress' utility from 'telemetry' container ...
    18:32:41 test_memory_checker.remove_stress_utilit L0092 INFO | 'stress' utility was removed.
    18:32:41 test_memory_checker.postcheck_critical_p L0172 INFO | Checking running status of critical processes in 'telemetry' container ...
    18:32:44 sonic.critical_process_status L0456 INFO | ====== supervisor process status for service telemetry ======
    18:32:44 test_memory_checker.postcheck_critical_p L0178 INFO | All critical processes in 'telemetry' container are running.
    PASSED [100%]
    ------------------------------ live log teardown -------------------------------
    18:32:44 test_memory_checker.modify_monit_config_ L0050 INFO | Restore original Monit configuration ...
    18:32:45 test_memory_checker.modify_monit_config_ L0053 INFO | Restart Monit service ...
    18:32:46 __init__.sanity_check L0249 INFO | No post-test check is required. Done post-test sanity check

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
N/A

### CODE REVIEW

File: tests/memory_checker/test_memory_checker.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def eatup_memory(duthost, container_name):
+    """Eats up more than 400MB memory in specified container.
+
+    Args:
+        duthost: The AnsibleHost object of DuT.
+        container_name: Name of container.
+
+    Retuerns:
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:102]: Typo: Returns

File: tests/memory_checker/test_memory_checker.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    """
+    logger.info("Removing 'stress' utility on device ...")
+    remove_cmd_result = duthost.shell("docker exec {} apt-get remove stress -y".format(container_name))
+    exit_code = remove_cmd_result["rc"]
+    pytest_assert(exit_code == 0, "Failed to remove 'stress' utility!")
+    logger.info("'stress' utility was removed.")
+
+
+def eatup_memory(duthost, container_name):
+    """Eats up more than 400MB memory in specified container.
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:96]: Can/should we make the amount of memory to consume configurable, like via passing a parameter? We could then extend this test to other containers besides telemetry.

File: tests/memory_checker/test_memory_checker.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        None.
+    """
+    logger.info("Removing 'stress' utility on device ...")
+    remove_cmd_result = duthost.shell("docker exec {} apt-get remove stress -y".format(container_name))
+    exit_code = remove_cmd_result["rc"]
+    pytest_assert(exit_code == 0, "Failed to remove 'stress' utility!")
+    logger.info("'stress' utility was removed.")
+
+
+def eatup_memory(duthost, container_name):
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:95]: Suggest s/eatup/consume/

File: tests/memory_checker/test_memory_checker.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        container_name: Name of container.
+
+    Retuerns:
+        None.
+    """
+    logger.info("Executing command 'stress -m 4' in '{}' container ...".format(container_name))
+    duthost.shell("docker exec {} stress -m 4".format(container_name), module_ignore_errors=True)
+
+
+def eatup_memory_and_restart_container(duthost, container_name):
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:109]: Suggest s/eatup/consume/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/memory_checker/test_memory_checker.py (hunk 2 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py:96]: Very good suggestion! Currently at here the parameter `-m` of `stress` utility is hard coded for telemetry container. What I am thinking is that we can configure this parameter according to the memory threshold value of each container. But let me update to pass this parameter as an argument of this function.

File: tests/memory_checker/test_memory_checker.py (hunk 4 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py]: Updated!

File: tests/memory_checker/test_memory_checker.py (hunk 1 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py]: Fixed.

File: tests/memory_checker/test_memory_checker.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+    Returns:
+        None.
+    """
+    logger.info("Executing command 'stress -m {}' in '{}' container ...".format(vm_workers, container_name))
+    duthost.shell("docker exec {} stress -m {}".format(container_name, vm_workers), module_ignore_errors=True)
+
+
+def consume_memory_and_restart_container(duthost, container_name, vm_workers):
+    """Invokes the 'stress' utility to consume memory more than threshold asynchronously and checkes
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:112]: Typo: s/checkes/checks/

File: tests/memory_checker/test_memory_checker.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    thread_pool.apply_async(consume_memory, (duthost, container_name, vm_workers))
+
+    logger.info("Waiting '{}' container to be stopped ...".format(container_name))
+    stopped = wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+                         CONTAINER_CHECK_INTERVAL_SECS,
+                         check_container_state, duthost, container_name, False)
+    pytest_assert(stopped, "Failed to stop '{}' container!".format(container_name))
+    logger.info("'{}' container is stopped.".format(container_name))
+
+    logger.info("Waiting '{}' container to be restarted ...".format(container_name))
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:136]: s/Waiting/Waiting for/

File: tests/memory_checker/test_memory_checker.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+    Returns:
+        None.
+
+    """
+    thread_pool = ThreadPool()
+
+    thread_pool.apply_async(consume_memory, (duthost, container_name, vm_workers))
+
+    logger.info("Waiting '{}' container to be stopped ...".format(container_name))
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:129]: s/Waiting/Waiting for/

File: tests/memory_checker/test_memory_checker.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def consume_memory_and_restart_container(duthost, container_name, vm_workers):
+    """Invokes the 'stress' utility to consume memory more than threshold asynchronously and checkes
+    whether the container can be stopped and restarted.
+
+    Args:
+        duthost: The AnsibleHost object of DuT.
+        container_name: Name of container.
+        vm_workers: Nnumber of workers which does the spinning on malloc()/free()
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:118]: s/Nnumber/Number/

File: tests/memory_checker/test_memory_checker.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    logger.info("'stress' utility was removed.")
+
+
+def consume_memory(duthost, container_name, vm_workers):
+    """Consumes memory more than the threshold value of specified container.
+
+    Args:
+        duthost: The AnsibleHost object of DuT.
+        container_name: Name of container.
+        vm_workers: Nnumber of workers which does the spinning on malloc()/free()
```
Comment (jleveque) [inline tests/memory_checker/test_memory_checker.py:101]: s/Nnumber/Number/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/memory_checker/test_memory_checker.py (hunk 5 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py]: Fixed.

File: tests/memory_checker/test_memory_checker.py (hunk 6 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py]: Good suggestion! Fixed.

**Review by jleveque — APPROVED** [review]

**Pull request merged**